boot/bootutil/src/loader.c

```diff
@@ -1512,88 +1558,433 @@ boot_set_image_ok(void)
+
+    /* Traverse through all of the TLVs to find the dependency TLVs. */
+    for (; off < end; off += sizeof(tlv) + tlv.it_len) {
+        rc = flash_area_read(fap, off, &tlv, sizeof(tlv));
+        if (rc != 0) {
+             rc = BOOT_EFLASH;
+             goto done;
+         }
+
+        if (tlv.it_type == IMAGE_TLV_DEPENDENCY) {
+            if (!dep_tlvs_found) {
+                dep_tlvs_found = true;
+            }
+
+            if (tlv.it_len != sizeof(dep)) {
+                rc = -1;
+                goto done;
+            }
+
+            rc = flash_area_read(fap, off + sizeof(tlv), &dep, tlv.it_len);
+            if (rc != 0) {
+                rc = BOOT_EFLASH;
+                goto done;
+            }
+
+            /* Verify dependency and modify the swap type if not satisfied. */
+            rc = boot_verify_single_dependency(&dep);
+            if (rc != 0) {
+                /* Dependency not satisfied. */
+                rc = -1;
+                goto done;
+            }
+
+            /* Dependency satisfied, no action needed.
+             * Continue with the next TLV entry.
+             */
+        } else if (dep_tlvs_found) {
+            /* The dependency TLVs are contiguous in the TLV area. If a
+             * dependency had already been found and the last read TLV
+             * has a different type then there are no more dependency TLVs.
+             * The search can be finished.
+             */
+            break;
+        }
+    }
+
+done:
+    flash_area_close(fap);
+    return rc;
+}
+
+/**
+ * Verify whether the image dependencies in the TLV area are
+ * all satisfied and modify the swap type if necessary.
+ *
+ * @return                  0 if all dependencies are satisfied,
+ *                          nonzero otherwise.
+ */
+static int
+boot_verify_single_image_dependency(void)
+{
+    size_t slot;
+    int32_t rc;
+
+    /* Determine the source of the dependency TLVs. Those dependencies have to
+     * be checked which belong to the image that will be located in the primary
+     * slot after the firmware update process.
+     */
+    if (boot_data->swap_type != BOOT_SWAP_TYPE_NONE &&
+        boot_data->swap_type != BOOT_SWAP_TYPE_FAIL) {
+        slot = BOOT_SECONDARY_SLOT;
+    } else {
+        slot = BOOT_PRIMARY_SLOT;
+    }
+
+    rc = boot_verify_all_dependency(slot);
+
+    return rc;
+}
+
+/**
+ * Iterate over all the images and verify whether the image dependencies in the
+ * TLV area are all satisfied and update the related swap type if necessary.
+ */
+static void
+boot_verify_all_image_dependency(void)
+{
+    size_t img_cnt = 0;
+
+    while(img_cnt < BOOT_IMAGE_NUMBER) {
+        boot_set_active_image(img_cnt);
+
+        if (boot_verify_single_image_dependency() == 0) {
+            /* All dependencies've been satisfied, continue with next image. */
+            img_cnt++;
+        } else {
+            /* Dependency check needs to be restarted. */
+            img_cnt = 0;
```

Comment:
The only time I've ever hit this line, it just causes this code to loop forever. I don't think this code is right, and will probably just loop forever if the dependencies haven't been met. I have yet to write tests explicitly for this, but by just having a SUIT manifest, it is easy to get this to loop.

Reply:
There is always at least 1 valid configuration, it is assumed that the initial state is consistent. In the worst case, the system returns to this initial state after dependency check.
The logic in the boot_is_version_sufficient() function was wrong... I've updated it. I guess this is why it stucked in a loop.